### PR TITLE
Amazon Web App Tests

### PR DESCRIPTION
**Pull request by edgaremirborbonjimenez**
Login Test
Logout Test
Login with wrong info
Check Product info
Add to the Cart
Remove from the cart

### CODE REVIEW

File: src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/CartProduct.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package com.zebrunner.carina.demo.gui.components.mycomponents;
+
+import com.zebrunner.carina.webdriver.decorator.ExtendedWebElement;
+import com.zebrunner.carina.webdriver.gui.AbstractUIObject;
+import org.openqa.selenium.SearchContext;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.FindBy;
+
+public class CartProduct extends AbstractUIObject {
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/CartProduct.java:9]: It is more beneficial to make all these methods generic and move them into the Card class. From this class you can make a class to store data for the CartProduct object.

File: src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    private ExtendedWebElement yourAccountOption;
+
+    @FindBy(xpath = "//input[@id=\"twotabsearchtextbox\"]")
+    private ExtendedWebElement searchTextBox;
+
+    public Header(WebDriver driver, SearchContext searchContext){
+        super(driver,searchContext);
+    }
+
+    public void selectYourAccountOption(String option){
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java:39]: It is beneficial to pass Enum to this method

File: src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        this.yourAccountOption = this.yourAccountOption.format(option);
+        WebDriverWait wait = new WebDriverWait(getDriver(), Duration.ofSeconds(waitingTime));
+        wait.until(ExpectedConditions.visibilityOf(this.yourAccountOption));
+        this.yourAccountOption.click();
+    }
+
+    public void hoverNavToolsOption(String navToolOption){
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java:46]: It is beneficial to pass Enum to this method

File: src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this.navToolsOptions.format(navToolOption).hover();
+    }
+
+    public String getWelcomeMessage(){
+        return welcomeMessage.getText();
+    }
+
+    public void searchProduct(String productName){
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/Header.java:54]: It is possible to return a page from this method

File: src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/ProductItem.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.zebrunner.carina.demo.gui.pages.mypages.AmazonProductPage;
+import com.zebrunner.carina.webdriver.decorator.ExtendedWebElement;
+import com.zebrunner.carina.webdriver.gui.AbstractUIObject;
+import com.zebrunner.carina.webdriver.locator.Context;
+import org.openqa.selenium.SearchContext;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.FindBy;
+
+public class ProductItem extends AbstractUIObject {
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/components/mycomponents/ProductItem.java:11]: It doesn't make sense to describe this class as a component if you don't use it on multiple pages

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonCartPage.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    private List<CartProduct> products;
+
+    public AmazonCartPage(WebDriver driver){
+        super(driver);
+        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(this.waitingTime));
+        wait.until(ExpectedConditions.visibilityOf(products.get(0)));
+    }
+
+    public Double getSubTotal(){
+        return Double.parseDouble(subTotal.getText().substring(1,subTotal.getText().length()));
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonCartPage.java:29]: It's better to handle the exception here

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithSession.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        this.header.hoverNavToolsOption("nav-link-accountList");
+    }
+
+    public AmazonSignInPage signOut(){
+        header.selectYourAccountOption("Sign Out");
+        return  new AmazonSignInPage(getDriver());
+    }
+
+    public String getSignOutText(){
+        return this.yourAccountOption.format("Sign Out").getText();
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithSession.java:32]: It is not correct to search for the text "Sign Out" using a locator that contains the text "Sign Out". It is more correct in this case to check that the element is present.

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithSession.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Override
+    public boolean isPageOpened() {
+        return getDriver().getCurrentUrl().contains("signin");
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithSession.java:37]: It's better to add setPageOpeningStrategy(PageOpeningStrategy.BY_ELEMENT) or setPageAbsoluteURL("") to the class constructor and use the existing isPageOpened() method

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithoutSession.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public AmazonSignInPage goToSignPage(){
+        signInOption.click();
+        return new AmazonSignInPage(getDriver());
+    }
+
+    @Override
+    public boolean isPageOpened() {
+        return getDriver().getCurrentUrl().contains("https://www.amazon.com/");
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonPageWithoutSession.java:26]: It's better to add setPageOpeningStrategy(PageOpeningStrategy.BY_ELEMENT) or setPageAbsoluteURL("") to the class constructor and use the existing isPageOpened() method

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonProductPage.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    ExtendedWebElement addToCart;
+
+
+    public AmazonProductPage(WebDriver driver){
+        super(driver);
+        WebDriverWait wait = new WebDriverWait(getDriver(), Duration.ofSeconds(this.waitingTime));
+        wait.until(ExpectedConditions.visibilityOf(title));
+    }
+
+    public String getProductTitle(){
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonProductPage.java:44]: The best option is to have a method in this class to return an object with all the data, then have a method in the CartPage that returns the same object with data and compare the objects in the test

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonSignInPage.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import com.zebrunner.carina.webdriver.decorator.ExtendedWebElement;
+import com.zebrunner.carina.webdriver.gui.AbstractPage;
+import com.zebrunner.carina.webdriver.locator.Context;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.FindBy;
+
+public class AmazonSignInPage extends AbstractPage {
+
+    private static String signInUrl = "https://www.amazon.com/ap/signin?openid.pape.max_auth_age=0&openid.return_to=https%3A%2F%2Fwww.amazon.com%2Fgp%2Fcss%2Fhomepage.html%3Fref_%3Dnav_ya_signin&openid.identity=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.assoc_handle=usflex&openid.mode=checkid_setup&openid.claimed_id=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0%2Fidentifier_select&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0";
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonSignInPage.java:12]: I think it’s better to store this in property files

File: src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonSignInPage.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    public AmazonSignInPage(WebDriver driver){
+        super(driver);
+    }
+
+    public String getTitle(){
+        return getDriver().findElement(By.tagName("h1")).getText();
+    }
+
+    public AmazonPageWithSession signIn(String emailOrPhone, String password){
```
Comment (aostapenko92) [inline src/main/java/com/zebrunner/carina/demo/gui/pages/mypages/AmazonSignInPage.java:42]: It's good to pass a User object here

File: src/test/java/com/zebrunner/carina/demo/AmazonWebTest.java [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        softAssert.assertAll();
+    }
+
+    @Test(description = "Validating Sign out")
+    @MethodOwner(owner = "emir")
+    public void validSignOutTest(){
+        //Sign In
+        AmazonSignInPage signInPage = new AmazonSignInPage(getDriver());
+        signInPage.open();
+        AmazonPageWithSession homePageWithSession = signInPage.signIn("eeborbonjimenez@laba.solvd.com","eeborbonjimenez@laba.solvd.com");
```
Comment (aostapenko92) [inline src/test/java/com/zebrunner/carina/demo/AmazonWebTest.java:49]: User data should be stored in property files

File: src/test/java/com/zebrunner/carina/demo/WebSampleTest.java [hunk 14]
```diff
@@ -143,4 +147,56 @@ public void testBrandGroup() {
...
+
+        //Va a la pagina de celulares de Nokia
+        //driver.findElement(By.xpath("//*[@id=\"body\"]/aside/div[1]/ul/li[4]/a")).click();
+        //driver.findElement(By.xpath("//a[text()=\"Nokia\"]"));
+
+//        driver.findElement(By.cssSelector("#topsearch>input")).sendKeys("Huawei");
+//        driver.findElement(By.cssSelector("#topsearch>input")).sendKeys(Keys.ENTER);
+        homePage.searchInput("Huawei");
+
+        String searchText = driver.findElement(By.tagName("h1")).getText();
```
Comment (aostapenko92) [inline src/test/java/com/zebrunner/carina/demo/WebSampleTest.java:171]: It is not correct to work with locors in test methods

**Review by aostapenko92 — COMMENTED** [review]
I didn’t describe the same mistakes, you need to find them and fix them yourself

File: src/test/java/com/zebrunner/carina/demo/WebSampleTest.java (hunk 14 above)
Comment (edgaremirborbonjimenez) [reply src/test/java/com/zebrunner/carina/demo/WebSampleTest.java:171]: Oh this was the test from the theoretical class hehe, was an error add it to the PR

**Pull request merged**